### PR TITLE
The models now promise

### LINKED ISSUE #5: Evaluate Promises
Should we be using promises internally in here?  Since we're going to be interacting with multiple models.

### PR DESCRIPTION
**Pull request by FenrirUnbound**
Decorated all the models to have a Promise interface. The endpoints are backwards compatible: if they are given a callback, they will invoke the callback as expected. If you omit a callback function, you'll receive a promise instead.

Merging this PR will resolve #5

### CODE REVIEW

File: lib/nodeify.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    const args = Array.from(arguments);
+    const func = args.shift();
+    let callback = args[args.length - 1];
+
+    if (typeof callback === 'function') {
+        // To allow thrown errors to _NOT_ be caught in the promise chain
+        return process.nextTick(() => func.apply(null, args));
+    } else if (typeof callback === 'undefined') {
+        args.pop();
+    }
```
Comment (stjohnjohnson) [inline lib/nodeify.js]: What would an else be here?

**Comment by stjohnjohnson** [comment]
Okay, this is an awesome PR.  Super cool.  I would bump the minor rev on the package.json.

Quick question, is the intention to support both callback and promise for everything in models going forward?

File: lib/nodeify.js (hunk 1 above)
Comment (FenrirUnbound) [inline lib/nodeify.js]: Nothing. It would fall through the rest of the method.

**Comment by FenrirUnbound** [comment]
> is the intention to support both callback and promise for everything in models going forward?

For the near-term, yes. Going all-in with promises would require us to switch everything over from passing in callbacks to a Promise-chain.

**Comment by petey** [comment]
:shipit:

**Comment by nkatzman** [comment]
Squash commits and then add that comment in the nodeify code 👍

**Comment by FenrirUnbound** [comment]
Commits squashed.

> add that comment in the nodeify code 

https://github.com/screwdriver-cd/models/pull/35/files#diff-cb6fb0408e4bce588c86b3c137060f54R20 😸

**Comment by d2lam** [comment]
👍  Makes sense

**Pull request merged**